### PR TITLE
Tensorflow Dataloader

### PR DESCRIPTION
**Pull request by soumik12345**
TODOs:

- [ ] Parallelize the array generation using multi-threading.
- [ ] Randomly sample a `2^k` number of points and corresponding labels. This will be similar to resize for images.
- [ ] Normalize point cloud.
- [ ] Apply random noise/jitter.
- [ ] Apply random rotation along z-axis.
- [ ] Create `tf.data.Dataset` objects for train and validation set.

### CODE REVIEW

File: point_seg/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+from glob import glob
+from tqdm import tqdm
+import tensorflow as tf
+import plotly.express as px
+import matplotlib.pyplot as plt
+
+
+class ShapeNetCoreLoader:
+
+    def __init__(self, object_category: str = 'Airplane', n_sampled_points: int = 1024) -> None:
```
Comment (sayakpaul) [inline point_seg/dataloader.py:15]: Here, prefer adding a docstring since `n_sampled_points` arg can be easily confused.

File: point_seg/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                self.test_point_clouds.append(point_cloud)
+    
+    def visualize_data_plotly(self, index):
+        fig = px.scatter_3d(
+            pd.DataFrame(
+                data={
+                    'x': self.point_clouds[index][:, 0],
+                    'y': self.point_clouds[index][:, 1],
+                    'z': self.point_clouds[index][:, 2],
+                    'label': self.all_labels[index]
```
Comment (sayakpaul) [inline point_seg/dataloader.py:101]: Seems like this dataframe could be created in a class method and can be reused in the visualization utils.

File: point_seg/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        plt.show()
+    
+    def _load_data(self, point_cloud, label_cloud):
+        point_cloud.set_shape([self.n_sampled_points, 3])
+        label_cloud.set_shape([self.n_sampled_points, len(self.labels)])
+        return point_cloud, label_cloud
+    
+    def _generate_dataset(self, point_clouds, label_clouds, batch_size: int):
+        dataset = tf.data.Dataset.from_tensor_slices((point_clouds, label_clouds))
+        dataset = dataset.map(self._load_data, num_parallel_calls=tf.data.AUTOTUNE)
```
Comment (sayakpaul) [inline point_seg/dataloader.py:140]: Define `tf.data.AUTOTUNE` as a module-level constant.

File: point_seg/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        dataset = dataset.batch(batch_size=batch_size, drop_remainder=True)
```
Comment (sayakpaul) [inline point_seg/dataloader.py:141]: Why `drop_remainder=True`? Prefer adding that as a comment if the usage is non-trivial.

File: point_seg/dataloader.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                    label=label, alpha = 0.5, c=self.colors[index]
+                ) 
+            except IndexError:
+                pass
+        ax.legend()
+        plt.show()
+    
+    def _load_data(self, point_cloud, label_cloud):
+        point_cloud.set_shape([self.n_sampled_points, 3])
+        label_cloud.set_shape([self.n_sampled_points, len(self.labels)])
```
Comment (sayakpaul) [inline point_seg/dataloader.py:135]: What happens without these two?

**Review by sayakpaul — CHANGES_REQUESTED** [review]
@soumik12345 seems like `load_data()` needs to be parallelized? 

If so, I would like to have a crack at it later today.

File: point_seg/dataloader.py (hunk 4 above)
Comment (soumik12345) [reply point_seg/dataloader.py:141]: > @soumik12345 seems like `load_data()` needs to be parallelized?
> 
> If so, I would like to have a crack at it later today.
> 
> From the notebook:
> 
> * There seems to be something wrong with the indices:
> 
> ```python
> for _ in range(10):
>     i = random.randint(0, len(point_clouds) - 1)
>     print(f'point_clouds[{i}].shape:', point_clouds[0].shape)
>     print(f'point_cloud_labels[{i}].shape:', point_cloud_labels[0].shape)
>     print(f'all_labels[{i}][:5]:', all_labels[i][:5], '\n')
> ```
> 
> * For convenience, add headers before the code cells. You can also group together multiple cells and add a single header for them.

HI @sayakpaul , can you please specify what did you find wrong with the indices?

File: point_seg/dataloader.py (hunk 5 above)
Comment (soumik12345) [reply point_seg/dataloader.py:135]: `<BatchDataset shapes: ((16, 1024, 3), (16, 1024, 4)), types: (tf.float64, tf.float32)>` would appear as `<BatchDataset shapes: ((16, None, None), (16, None, None)), types: (tf.float64, tf.float32)>`.

File: point_seg/dataloader.py (hunk 1 above)
Comment (soumik12345) [reply point_seg/dataloader.py:22]: Added docstrings.

File: point_seg/dataloader.py (hunk 2 above)
Comment (soumik12345) [reply point_seg/dataloader.py]: Done.

File: point_seg/dataloader.py (hunk 4 above)
Comment (soumik12345) [reply point_seg/dataloader.py]: Removed `drop_remainder=True`. Don't think there's a need to drop the smaller remaining batch.

File: point_seg/dataloader.py (hunk 3 above)
Comment (soumik12345) [reply point_seg/dataloader.py:152]: @sayakpaul any particular reason?

Comment (sayakpaul) [reply point_seg/dataloader.py:152]: It's likely we will use it multiple times as we make progress with the rest of the utilities. But I think, for the time being, this suggestion can be ignored.

**Comment by sayakpaul** [comment]
@soumik12345 one doubt regarding:

> Apply random rotation along z-axis.

Why the z-axis in particular?

**Review by sayakpaul — APPROVED** [review]

**Comment by sayakpaul** [comment]
@soumik12345 I will try working on the rest of the point by tonight.

**Comment by soumik12345** [comment]
@sayakpaul , the dataloader notebook looks fine, I have 2 minor suggestions:

1. https://app.reviewnb.com/soumik12345/point-cloud-segmentation/blob/dataloader/notebooks%2Fdata_loader_multi_proc.ipynb/file/#comment-nb-02339837
2. I think that the notebook should be renamed to end-to-end tensorflow dataset creation (or something along similar lines).

I will proceed to integrate the e2e dataset creation with the `ShapeNetCoreLoader` class.

File: point_seg/dataloader/in_memory.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+
+class ShapeNetCoreLoaderInMemory:
+    """
+    In-memory Dataloader class for Shapenet Core Dataset.
+
+    Args:
+        object_category (str): One of the 12 objects from the ShapenetCore dataset.
+        n_sampled_points (int): Number of points to be sampled from each point cloud.
+    """
```
Comment (sayakpaul) [inline point_seg/dataloader/in_memory.py:20]: Since we have two modules for the dataset, it might be a good idea to specify their expected usage.

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+# Set Working Directory
+RUN mkdir /usr/src/point-cloud-segmentation
+WORKDIR /usr/src/point-cloud-segmentation
+
+# Set Environment Variables
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+
+COPY ./requirements.docker /usr/src/point-cloud-segmentation/requirements.docker
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply Dockerfile:12]: Added `requirements.docker`

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    libgtk2.0-dev \ 
+    python-numpy \ 
+    python-pycurl \ 
+    libatlas-base-dev \
+    gfortran \
+    webp \ 
+    python-opencv \ 
+    qt5-default \
+    libvtk6-dev \ 
+    zlib1g-dev
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply Dockerfile:37]: @sayakpaul On second thought, we actually don't need these, most of these are for opencv which we are not using.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+RUN pip install --upgrade pip setuptools wheel
+RUN pip install black==19.10b0 \
+    jupyter \
+    matplotlib==3.4.3 \
+    pandas==1.3.3 \
+    plotly==5.3.1 \
+    plotly-express==0.4.1 \
+    pytest==6.2.5 \
+    tqdm==4.62.3
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply Dockerfile:47]: @sayakpaul would be doing this from `requirements.docker`

File: point_seg/dataloader/in_memory.py (hunk 6 above)
Comment (soumik12345) [reply point_seg/dataloader/in_memory.py]: @sayakpaul elaborated the documentation on this.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+WORKDIR /usr/src/point-cloud-segmentation
+
+# Set Environment Variables
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+
+COPY ./requirements.docker /usr/src/point-cloud-segmentation/requirements.docker
+
+RUN pip install --upgrade pip setuptools wheel
+RUN pip install -r requirements.docker
```
[the remark this answers is not in the document]
Comment (soumik12345) [reply Dockerfile:15]: @sayakpaul I had kept `requirements.txt` around for setting the development environment. Basically, it contains the additional dependency of `tensorflow==2.6.0`.
